substratevm/src/com.oracle.svm.agent/src/com/oracle/svm/agent/BreakpointInterceptor.java

```diff
@@ -913,6 +918,52 @@ private static String asInternalSignature(Object paramTypesArray) {
         return null;
     }
 
+    private static boolean objectStreamClassConstructor(JNIEnvironment jni, Breakpoint bp) {
+        JNIObjectHandle callerClass = getDirectCallerClass();
+        JNIObjectHandle self = getObjectArgument(0);
+        JNIObjectHandle serializeTargetClass = getObjectArgument(1);
+        String serializeTargetClassName = getClassNameOrNull(jni, serializeTargetClass);
+        long checksum = 0;
+
+        JNIObjectHandle objectStreamClassInstance = newObjectL(jni, bp.clazz, bp.method, serializeTargetClass);
+        Object result = nullHandle().notEqual(objectStreamClassInstance);
+        if (clearException(jni)) {
+            result = false;
+        }
+        if (result.equals(true)) {
+            JNIObjectHandle cons = getObjectField(jni, bp.clazz, objectStreamClassInstance, "cons", "Ljava/lang/reflect/Constructor;");
+            String targetConstructorClassName = "";
+            if (nullHandle().notEqual(cons)) {
+                JNIObjectHandle constructorClazz = jniFunctions().getGetObjectClass().invoke(jni, cons);
+                try (CCharPointerHolder getDeclaringClassNameHolder = toCString("getDeclaringClass");
+                                CCharPointerHolder getDeclaringClassSigHolder = toCString("()Ljava/lang/Class;");) {
+                    JNIMethodId getDeclaringClassMId = jniFunctions().getGetMethodID().invoke(jni, constructorClazz, getDeclaringClassNameHolder.get(), getDeclaringClassSigHolder.get());
+                    targetConstructorClassName = getClassNameOrNull(jni, callObjectMethod(jni, cons, getDeclaringClassMId));
+                    checksum = targetConstructorClassName.hashCode();
```

Comment:
The constructor's declaring class is not just the superclass. It is the first unserializable super class. The hashcode is enough to tell whether the constructor's declaring class has been changed which may happen if the class' hierarchy is changed among different classloaders.

I examined the source code of JDK's _`MethodAccessorGenerator.generateSerializationConstructor(Class<?> declaringClass, Class<?>[] parameterTypes, Class<?>[] checkedExceptions, int modifiers, Class<?> targetConstructorClass)`_ and found the returned `SerializationConstructorAccessor` instance is determined only by the names of `declaringClass`(serialization target class) and `targetConstructorClass`(the first unserializable super class). 

For example, the following bytecode is from the `SerializationConstructorAccessor` class generated for `java.lang.String`. Only instructor 0 and 24 are differed from classes. Instruction 0 new `declaringClass` and instruction 24 invoke `targetConstructorClass`'s non-parameter-constructor.
```
 public java.lang.Object newInstance(java.lang.Object[]) throws java.lang.reflect.InvocationTargetException;
    descriptor: ([Ljava/lang/Object;)Ljava/lang/Object;
    flags: ACC_PUBLIC
    Code:
      stack=6, locals=2, args_size=2
         0: new           #6                  // class java/lang/String
         3: dup
         4: aload_1
         5: ifnull        24
         8: aload_1
         9: arraylength
        10: sipush        0
        13: if_icmpeq     24
        16: new           #22                 // class java/lang/IllegalArgumentException
        19: dup
        20: invokespecial #29                 // Method java/lang/IllegalArgumentException."<init>":()V
        23: athrow
        24: invokespecial #12                 // Method java/lang/Object."<init>":()V
        27: areturn
        28: invokespecial #42                 // Method java/lang/Object.toString:()Ljava/lang/String;
        31: new           #22                 // class java/lang/IllegalArgumentException
        34: dup_x1
        35: swap
        36: invokespecial #32                 // Method java/lang/IllegalArgumentException."<init>":(Ljava/lang/String;)V
        39: athrow
        40: new           #24                 // class java/lang/reflect/InvocationTargetException
        43: dup_x1
        44: swap
        45: invokespecial #35                 // Method java/lang/reflect/InvocationTargetException."<init>":(Ljava/lang/Throwable;)V
        48: athrow

```
While `computeDefaultSUID` calculates a checksum based on nearly every thing of `declaringClass`, so we should not use it to match the cached `SerializationConstructorAccessor`.

Reply:
Thanks for explaining. So we can tell whether the constructor that we generate during the image build matches the constructor from the agent run, but we cannot tell whether the serializable/deserializable class we have for the image build is the same as the ones seen during the agent run, unless their first unserializable superclass is different.
Is this really sufficient? Why not just use a computed SUID and ensure that we have an exact match? I'm wondering about the specific use case here, and also worry about extensibility of the configuration in case we want to include more per-class information in the future and want to ensure an exact match for that.

Reply:
I have applied the computed SUID. Not only the serialization target class itself's SUID is needed, all its serializable superclasses' SUIDs are also needed.